import sys
import os
import json
import time
from datetime import datetime
from PyQt5.QtWidgets import (QApplication, QMainWindow, QTabWidget, 
                             QWidget, QVBoxLayout, QFileDialog)  # è¡¥å……QFileDialogå¯¼å…¥
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QFont, QPixmap
from PyQt5.QtWidgets import QSizePolicy
from ui_forms.ui_chat_interface import Ui_chat_interface as ChatUi
from ui_forms.ui_image_generator import Ui_MainWindow as ImageUi
from ui_forms.ui_history_manager import Ui_MainWindow as HistoryUi

import threading
from dotenv import load_dotenv  # åŠ è½½.envæ–‡ä»¶
from openai import OpenAI
from PyQt5.QtWidgets import QMessageBox 
from PyQt5.QtWidgets import QTextEdit
import torch
from diffusers import StableDiffusionPipeline, DPMSolverMultistepScheduler
from PyQt5.QtCore import QThread, pyqtSignal



load_dotenv()



# åˆå§‹åŒ–OpenAIå®¢æˆ·ç«¯
openai_client = OpenAI(
    # å¦‚æœæ²¡æœ‰é…ç½®ç¯å¢ƒå˜é‡ï¼Œè¯·ç”¨é˜¿é‡Œäº‘ç™¾ç‚¼API Keyæ›¿æ¢ï¼šapi_key="sk-xxx"
    api_key=os.getenv("DASHSCOPE_API_KEY"),
    base_url="https://dashscope.aliyuncs.com/compatible-mode/v1",
)


class ChatInterface(QMainWindow):
    def __init__(self):
        super().__init__()
        self.ui = ChatUi()
        self.ui.setupUi(self)
        self.ui.centralwidget.setLayout(self.ui.verticalLayout)
        self.ui.verticalLayout.setContentsMargins(10, 10, 10, 10)
        self.ui.verticalLayout.setSpacing(10)
        # ç¦æ­¢èŠå¤©è®°å½•æ¡†ç¼–è¾‘
        self.ui.chat_history.setReadOnly(True)

        # æ·»åŠ ç‚¹å‡»æ¸…ç©ºè¾“å…¥æ¡†åŠŸèƒ½
        self.ui.text_input.mousePressEvent = self.on_input_click

        self.setup_connections()
    
    def on_input_click(self, event):
        """ç‚¹å‡»è¾“å…¥æ¡†æ—¶æ¸…ç©ºå†…å®¹"""
        if self.ui.text_input.toPlainText().strip() == "è¾“å…¥æ¶ˆæ¯...":
            self.ui.text_input.clear()
        # ç¡®ä¿æ­£å¸¸çš„ç‚¹å‡»äº‹ä»¶ç»§ç»­æ‰§è¡Œ
        super(type(self.ui.text_input), self.ui.text_input).mousePressEvent(event)

    def setup_connections(self):
        # ç»‘å®šæŒ‰é’®äº‹ä»¶
        self.ui.send_btn.clicked.connect(self.send_message)
        self.ui.clear_btn.clicked.connect(self.clear_chat)
        self.ui.text_input.keyPressEvent = self.on_text_input_key_press

    def on_text_input_key_press(self, event):
        """Enterå‘é€æ¶ˆæ¯ï¼ŒCtrl+Enteræ¢è¡Œï¼Œä¿ç•™å…¶ä»–æŒ‰é”®é»˜è®¤è¡Œä¸º"""
        if event.key() == Qt.Key_Return or event.key() == Qt.Key_Enter:
            if event.modifiers() == Qt.ControlModifier:
                # Ctrl+Enterï¼šæ’å…¥æ¢è¡Œ
                self.ui.text_input.insertPlainText("\n")
            else:
                # Enterï¼šå‘é€æ¶ˆæ¯
                self.send_message()
            event.accept()
        else:
            # å…¶ä»–æŒ‰é”®ï¼ˆå¦‚Backspaceï¼‰ä½¿ç”¨é»˜è®¤è¡Œä¸º
            QTextEdit.keyPressEvent(self.ui.text_input, event)

        
    def clear_chat(self):
        self.ui.chat_history.clear()

    def send_message(self):
        user_msg = self.ui.text_input.toPlainText().strip()
        if not user_msg:
            QMessageBox.warning(self, "è¾“å…¥æç¤º", "è¯·è¾“å…¥æ¶ˆæ¯å†…å®¹åå†å‘é€ï¼")
            return
        # 1. æ˜¾ç¤ºç”¨æˆ·æ¶ˆæ¯
        self.ui.chat_history.append(f"<b>[{self.get_current_time()}]</b> ä½ : {user_msg}")
        self.ui.text_input.clear()
        # ç¦ç”¨å‘é€æŒ‰é’®ï¼Œé˜²æ­¢é‡å¤å‘é€
        self.ui.send_btn.setEnabled(False)
        # æ»šåŠ¨åˆ°åº•éƒ¨
        self.ui.chat_history.verticalScrollBar().setValue(
            self.ui.chat_history.verticalScrollBar().maximum()
        )

        # 2. å¤šçº¿ç¨‹è°ƒç”¨OpenAI APIï¼ˆé¿å…UIå¡é¡¿ï¼‰
        threading.Thread(
            target=self.call_openai_api,
            args=(user_msg,),
            daemon=True  # çº¿ç¨‹éšä¸»ç¨‹åºé€€å‡ºè€Œå…³é—­
        ).start()

    def call_openai_api(self, user_msg):
        """è°ƒç”¨OpenAI APIè·å–å›å¤ï¼ˆå­çº¿ç¨‹ä¸­æ‰§è¡Œï¼‰"""
        # æ˜¾ç¤ºâ€œæ€è€ƒä¸­â€æç¤ºï¼ˆéœ€ç”¨QMetaObject.invokeMethodåˆ‡æ¢åˆ°ä¸»çº¿ç¨‹æ›´æ–°UIï¼‰
        self.update_chat_history("<i>[AIæ­£åœ¨æ€è€ƒ...]</i>")

        try:
            # è°ƒç”¨OpenAI ChatCompletion APIï¼ˆgpt-3.5-turboï¼Œæ€§ä»·æ¯”é«˜ï¼‰
            response = openai_client.chat.completions.create(
                model="deepseek-v3.2-exp",
                messages=[{"role": "user", "content": user_msg}],
                # é€šè¿‡ extra_body è®¾ç½® enable_thinking å¼€å¯æ€è€ƒæ¨¡å¼ï¼Œè¯¥å‚æ•°ä»…å¯¹ deepseek-v3.2-exp å’Œ deepseek-v3.1 æœ‰æ•ˆã€‚deepseek-v3 å’Œ deepseek-r1 è®¾å®šä¸ä¼šæŠ¥é”™
                extra_body={"enable_thinking": True},
                stream=True,
                stream_options={
                    "include_usage": True
                },
            )
            #ai_msg = response.choices[0].message.content.strip()

            ai_msg = ""
            for chunk in response:  # éå†æ¯ä¸ªæµå¼ç‰‡æ®µ
                if chunk.choices and chunk.choices[0].delta.content:
                    ai_msg += chunk.choices[0].delta.content  # ç´¯åŠ å†…å®¹

            # æ˜¾ç¤ºå®Œæ•´å›å¤
            self.update_chat_history(f"<b>[{self.get_current_time()}]</b> AI: {ai_msg.strip()}")
            self.save_chat_to_history(user_msg, ai_msg.strip())


        except Exception as e:
            # å¼‚å¸¸å¤„ç†ï¼ˆç½‘ç»œé”™è¯¯ã€å¯†é’¥é”™è¯¯ç­‰ï¼‰
            error_msg = f"AIå›å¤å¤±è´¥ï¼š{str(e)}"
            self.update_chat_history(f"<span style='color:red;'>{error_msg}</span>")
            QMessageBox.critical(self, "APIé”™è¯¯", error_msg)

        finally:
            # æ¢å¤å‘é€æŒ‰é’®å¯ç”¨
            self.ui.send_btn.setEnabled(True)

    def update_chat_history(self, content):
        """æ›´æ–°èŠå¤©è®°å½•ï¼ˆéœ€åœ¨ä¸»çº¿ç¨‹æ‰§è¡Œï¼Œé¿å…UIçº¿ç¨‹å®‰å…¨é—®é¢˜ï¼‰"""
        from PyQt5.QtCore import QMetaObject, Qt, Q_ARG
        QMetaObject.invokeMethod(
            self.ui.chat_history,
            "append",
            Qt.QueuedConnection,
            Q_ARG(str, content)
        )
        # æ»šåŠ¨åˆ°åº•éƒ¨ï¼ˆåŒæ ·éœ€ä¸»çº¿ç¨‹æ‰§è¡Œï¼‰
        QMetaObject.invokeMethod(
            self.ui.chat_history.verticalScrollBar(),
            "setValue",
            Qt.QueuedConnection,
            Q_ARG(int, self.ui.chat_history.verticalScrollBar().maximum())
        )

    def save_chat_to_history(self, user_msg, ai_msg):
        """ä¿å­˜èŠå¤©è®°å½•ï¼ˆå«å®Œæ•´ä¸Šä¸‹æ–‡ï¼Œæ”¯æŒç»­èŠï¼‰"""
        try:
            # å®šä¹‰å†å²è®°å½•è·¯å¾„ï¼ˆä½¿ç”¨ç»å¯¹è·¯å¾„ï¼Œé¿å…ç›¸å¯¹è·¯å¾„é—®é¢˜ï¼‰
            history_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "../history.json"))
            print(f"ä¿å­˜å†å²åˆ°ï¼š{history_path}")  # è°ƒè¯•æ—¥å¿—
            
            # è¯»å–ç°æœ‰å†å²ï¼ˆæ–‡ä»¶ä¸å­˜åœ¨åˆ™åˆå§‹åŒ–ï¼‰
            history = []
            if os.path.exists(history_path):
                with open(history_path, "r", encoding="utf-8") as f:
                    history = json.load(f)
            
            # æ„é€ å®Œæ•´å¯¹è¯ä¸Šä¸‹æ–‡ï¼ˆç”¨äºç»­èŠï¼‰
            chat_context = [
                {"role": "user", "content": user_msg},
                {"role": "assistant", "content": ai_msg}
            ]
            
            # æ·»åŠ æ–°è®°å½•ï¼ˆåŒ…å«æ‘˜è¦å’Œä¸Šä¸‹æ–‡ï¼‰
            history.append({
                "id": str(uuid.uuid4()),  # å”¯ä¸€æ ‡è¯†ï¼Œç”¨äºç»­èŠ
                "time": self.get_current_time(),
                "type": "chat",
                "summary": f"ä½ : {user_msg[:30]}...\nAI: {ai_msg[:30]}...",  # æ‘˜è¦
                "context": chat_context  # å®Œæ•´ä¸Šä¸‹æ–‡
            })
            
            # å†™å…¥æ–‡ä»¶ï¼ˆç¡®ä¿ç›®å½•å­˜åœ¨ï¼‰
            os.makedirs(os.path.dirname(history_path), exist_ok=True)
            with open(history_path, "w", encoding="utf-8") as f:
                json.dump(history, f, ensure_ascii=False, indent=2)
            
            print("èŠå¤©è®°å½•ä¿å­˜æˆåŠŸ")  # è°ƒè¯•æ—¥å¿—
            
        except Exception as e:
            error_msg = f"ä¿å­˜èŠå¤©å†å²å¤±è´¥ï¼š{str(e)}"
            print(error_msg)  # æ‰“å°é”™è¯¯æ—¥å¿—
            QMessageBox.warning(self, "ä¿å­˜å¤±è´¥", error_msg)

    @staticmethod
    def get_current_time():
        """è·å–å½“å‰æ—¶é—´ï¼ˆæ ¼å¼ï¼š2024-05-20 15:30:45ï¼‰"""
        return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

        
    # def get_ai_response(self, prompt):
    #     # æ¨¡æ‹ŸAIå›å¤ï¼Œå®é™…é¡¹ç›®ä¸­æ›¿æ¢ä¸ºçœŸå®è°ƒç”¨
    #     import time
    #     time.sleep(1)  # æ¨¡æ‹Ÿç½‘ç»œå»¶è¿Ÿ
    #     return f"è¿™æ˜¯å¯¹'{prompt}'çš„å›å¤ã€‚åœ¨å®é™…åº”ç”¨ä¸­ï¼Œè¿™é‡Œä¼šæ˜¾ç¤ºAIæ¨¡å‹ç”Ÿæˆçš„å†…å®¹ã€‚"

class ImageGenerateThread(QThread):
    """å›¾åƒç”Ÿæˆå­çº¿ç¨‹ï¼ˆé¿å…UIå¡é¡¿ï¼‰"""
    # ä¿¡å·ï¼šç”Ÿæˆè¿›åº¦ï¼ˆæ­¥æ•°ï¼‰ã€ç”Ÿæˆå®Œæˆï¼ˆå›¾åƒè·¯å¾„ï¼‰ã€ç”Ÿæˆå¤±è´¥ï¼ˆé”™è¯¯ä¿¡æ¯ï¼‰
    progress_signal = pyqtSignal(int)
    finish_signal = pyqtSignal(str)
    error_signal = pyqtSignal(str)

    def __init__(self, prompt, width, height, num_inference_steps):
        super().__init__()
        self.prompt = prompt
        self.width = width
        self.height = height
        self.num_inference_steps = num_inference_steps
        self.pipeline = None

    def run(self):
        """çº¿ç¨‹æ‰§è¡Œï¼šåŠ è½½æ¨¡å‹+ç”Ÿæˆå›¾åƒ"""
        try:
            # 1. åŠ è½½Stable Diffusionæ¨¡å‹ï¼ˆä¼˜å…ˆGPUï¼Œæ— GPUåˆ™ç”¨CPUï¼‰
            device = "cuda" if torch.cuda.is_available() else "cpu"
            # ä½¿ç”¨DPMSolverè°ƒåº¦å™¨ï¼Œç”Ÿæˆé€Ÿåº¦æ¯”é»˜è®¤å¿«
            scheduler = DPMSolverMultistepScheduler.from_pretrained(
                "runwayml/stable-diffusion-v1-5", subfolder="scheduler"
            )
            self.pipeline = StableDiffusionPipeline.from_pretrained(
                "runwayml/stable-diffusion-v1-5",
                scheduler=scheduler,
                torch_dtype=torch.float16 if device == "cuda" else torch.float32
            ).to(device)

            # 2. ç”Ÿæˆå›¾åƒï¼ˆå¸¦è¿›åº¦åé¦ˆï¼‰
            def progress_callback(step, timestep, latents):
                """è¿›åº¦å›è°ƒå‡½æ•°ï¼Œå‘é€å½“å‰æ­¥æ•°"""
                self.progress_signal.emit(step + 1)  # stepä»0å¼€å§‹ï¼Œ+1åä¸æ€»æ­¥æ•°å¯¹åº”

            image = self.pipeline(
                prompt=self.prompt,
                width=self.width,
                height=self.height,
                num_inference_steps=self.num_inference_steps,
                callback=progress_callback  # ç»‘å®šè¿›åº¦å›è°ƒ
            ).images[0]

            # 3. ä¿å­˜å›¾åƒåˆ°resourcesç›®å½•
            output_path = os.path.join("resources", f"sd_output_{self.get_current_time()}.png")
            image.save(output_path)
            self.finish_signal.emit(output_path)

        except Exception as e:
            self.error_signal.emit(str(e))

    @staticmethod
    def get_current_time():
        """ç”Ÿæˆå”¯ä¸€æ–‡ä»¶åï¼ˆé¿å…é‡å¤ï¼‰"""
        return datetime.now().strftime("%Y%m%d%H%M%S")

class ImageGenerator(QMainWindow):
    def __init__(self):
        super().__init__()
        self.ui = ImageUi()
        self.ui.setupUi(self)
        # å¸ƒå±€é…ç½®
        self.ui.centralwidget.setLayout(self.ui.horizontalLayout)  # horizontalLayoutæ˜¯Designerä¸­ä¸­å¤®éƒ¨ä»¶çš„å¸ƒå±€å
        self.ui.horizontalLayout.setContentsMargins(10, 10, 10, 10)
        self.ui.horizontalLayout.setSpacing(10)
        # åˆå§‹åŒ–å˜é‡
        self.current_pixmap = None
        self.generate_thread = None  # å›¾åƒç”Ÿæˆçº¿ç¨‹
        # è®¾ç½®å‚æ•°èŒƒå›´
        self.ui.width_spin.setRange(256, 1024)    # å®½åº¦ï¼š256-1024
        self.ui.height_spin.setRange(256, 1024)   # é«˜åº¦ï¼š256-1024
        self.ui.steps_slider.setRange(10, 50)     # æ­¥æ•°ï¼š10-50
        self.ui.steps_slider.setValue(20)         # é»˜è®¤æ­¥æ•°ï¼š20
        self.ui.steps_label.setText(f"ç”Ÿæˆæ­¥æ•°ï¼š{20}")  # æ˜¾ç¤ºå½“å‰æ­¥æ•°ï¼ˆéœ€åœ¨Designerä¸­æ·»åŠ QLabelå‘½åä¸ºsteps_labelï¼‰
        self.setup_connections()

    def setup_connections(self):
        self.ui.generate_btn.clicked.connect(self.start_generate)
        self.ui.save_btn.clicked.connect(self.save_image)
        self.ui.steps_slider.valueChanged.connect(self.update_steps_label)

    def update_steps_label(self, value):
        """æ›´æ–°ç”Ÿæˆæ­¥æ•°æ˜¾ç¤º"""
        self.ui.steps_label.setText(f"ç”Ÿæˆæ­¥æ•°ï¼š{value}")

    def start_generate(self):
        """å¼€å§‹ç”Ÿæˆå›¾åƒï¼ˆå¯åŠ¨å­çº¿ç¨‹ï¼‰"""
        prompt = self.ui.prompt_input.text().strip()
        width = self.ui.width_spin.value()
        height = self.ui.height_spin.value()
        steps = self.ui.steps_slider.value()

        # è¾“å…¥æ ¡éªŒ
        if not prompt:
            QMessageBox.warning(self, "è¾“å…¥æç¤º", "è¯·è¾“å…¥å›¾åƒæè¿°ï¼ˆPromptï¼‰ï¼")
            return
        if self.generate_thread and self.generate_thread.isRunning():
            QMessageBox.information(self, "ç”Ÿæˆæç¤º", "å½“å‰å·²æœ‰ç”Ÿæˆä»»åŠ¡åœ¨è¿è¡Œï¼Œè¯·ç¨å€™ï¼")
            return

        # åˆå§‹åŒ–çŠ¶æ€
        self.ui.image_preview.clear()
        self.ui.image_preview.setText(f"æ­£åœ¨åŠ è½½æ¨¡å‹...ï¼ˆ{('GPUåŠ é€Ÿ' if torch.cuda.is_available() else 'CPU')}ï¼‰")
        self.ui.generate_btn.setEnabled(False)
        self.ui.save_btn.setEnabled(False)
        # é‡ç½®è¿›åº¦æ¡ï¼ˆéœ€åœ¨Designerä¸­æ·»åŠ QProgressBarå‘½åä¸ºgenerate_progressï¼‰
        self.ui.generate_progress.setRange(0, steps)
        self.ui.generate_progress.setValue(0)

        # å¯åŠ¨ç”Ÿæˆçº¿ç¨‹
        self.generate_thread = ImageGenerateThread(prompt, width, height, steps)
        self.generate_thread.progress_signal.connect(self.update_generate_progress)
        self.generate_thread.finish_signal.connect(self.on_generate_finish)
        self.generate_thread.error_signal.connect(self.on_generate_error)
        self.generate_thread.start()

    def update_generate_progress(self, step):
        """æ›´æ–°ç”Ÿæˆè¿›åº¦æ¡"""
        self.ui.generate_progress.setValue(step)
        self.ui.image_preview.setText(f"æ­£åœ¨ç”Ÿæˆï¼š{step}/{self.ui.steps_slider.value()} æ­¥")

    def on_generate_finish(self, image_path):
        """ç”Ÿæˆå®Œæˆï¼šæ˜¾ç¤ºå›¾åƒ"""
        self.current_pixmap = QPixmap(image_path)
        if not self.current_pixmap.isNull():
            # è‡ªé€‚åº”ç¼©æ”¾å›¾åƒ
            scaled_pixmap = self.current_pixmap.scaled(
                self.ui.image_preview.size(),
                Qt.KeepAspectRatio,
                Qt.SmoothTransformation
            )
            self.ui.image_preview.setPixmap(scaled_pixmap)
            # ä¿å­˜åˆ°å†å²è®°å½•
            self.save_image_to_history(image_path)
        else:
            self.ui.image_preview.setText("ç”Ÿæˆå¤±è´¥ï¼šæ— æ³•åŠ è½½å›¾åƒæ–‡ä»¶")

        # æ¢å¤æ§ä»¶çŠ¶æ€
        self.ui.generate_btn.setEnabled(True)
        self.ui.save_btn.setEnabled(True)

    def on_generate_error(self, error_msg):
        """ç”Ÿæˆå¤±è´¥ï¼šæ˜¾ç¤ºé”™è¯¯ä¿¡æ¯"""
        error_info = f"ç”Ÿæˆå¤±è´¥ï¼š{error_msg}"
        self.ui.image_preview.setText(error_info)
        QMessageBox.critical(self, "ç”Ÿæˆé”™è¯¯", error_info)
        # æ¢å¤æ§ä»¶çŠ¶æ€
        self.ui.generate_btn.setEnabled(True)

    def save_image(self):
        """ä¿å­˜ç”Ÿæˆçš„å›¾åƒ"""
        if not self.current_pixmap or self.current_pixmap.isNull():
            QMessageBox.warning(self, "ä¿å­˜æç¤º", "æš‚æ— ç”Ÿæˆçš„å›¾åƒå¯ä¿å­˜ï¼")
            return

        # å¼¹å‡ºä¿å­˜å¯¹è¯æ¡†
        file_path, _ = QFileDialog.getSaveFileName(
            self, "ä¿å­˜å›¾åƒ",
            f"sd_output_{self.get_current_time()}.png",
            "PNGæ–‡ä»¶ (*.png);;JPEGæ–‡ä»¶ (*.jpg);;æ‰€æœ‰æ–‡ä»¶ (*.*)"
        )
        if file_path:
            if self.current_pixmap.save(file_path):
                QMessageBox.information(self, "ä¿å­˜æˆåŠŸ", f"å›¾åƒå·²ä¿å­˜åˆ°ï¼š\n{file_path}")
            else:
                QMessageBox.error(self, "ä¿å­˜å¤±è´¥", "æ— æ³•ä¿å­˜å›¾åƒï¼Œè¯·æ£€æŸ¥è·¯å¾„æƒé™ï¼")

    def save_image_to_history(self, image_path):
        """ä¿å­˜å›¾åƒç”Ÿæˆè®°å½•åˆ°history.json"""
        try:
            history = []
            if os.path.exists("history.json"):
                with open("history.json", "r", encoding="utf-8") as f:
                    history = json.load(f)
            # æ·»åŠ è®°å½•ï¼ˆåŒ…å«å›¾åƒè·¯å¾„ï¼Œæ–¹ä¾¿åç»­æŸ¥çœ‹ï¼‰
            history.append({
                "time": self.get_current_time(),
                "type": "image",
                "content": f"æè¿°ï¼š{self.ui.prompt_input.text().strip()}\nå°ºå¯¸ï¼š{self.ui.width_spin.value()}x{self.ui.height_spin.value()}\nè·¯å¾„ï¼š{image_path}"
            })
            # å†™å…¥æ–‡ä»¶
            with open("history.json", "w", encoding="utf-8") as f:
                json.dump(history, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"ä¿å­˜å›¾åƒå†å²å¤±è´¥ï¼š{e}")

    @staticmethod
    def get_current_time():
        """è·å–å½“å‰æ—¶é—´ï¼ˆæ ¼å¼ï¼š2024-05-20 15:30:45ï¼‰"""
        return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    def resizeEvent(self, event):
        """çª—å£ç¼©æ”¾æ—¶ï¼Œæ›´æ–°å›¾åƒé¢„è§ˆ"""
        super().resizeEvent(event)
        if self.current_pixmap and not self.current_pixmap.isNull():
            self.update_image_preview()

    def update_image_preview(self):
        """æ›´æ–°å›¾åƒé¢„è§ˆï¼ˆè‡ªé€‚åº”çª—å£å¤§å°ï¼‰"""
        scaled_pixmap = self.current_pixmap.scaled(
            self.ui.image_preview.size(),
            Qt.KeepAspectRatio,
            Qt.SmoothTransformation
        )
        self.ui.image_preview.setPixmap(scaled_pixmap)

class HistoryManager(QMainWindow):
    def __init__(self):
        super().__init__()
        self.ui = HistoryUi()
        self.ui.setupUi(self)
        # å¸ƒå±€é…ç½®
        self.ui.centralwidget.setLayout(self.ui.horizontalLayout)  # horizontalLayoutæ˜¯Designerä¸­ä¸­å¤®éƒ¨ä»¶çš„å¸ƒå±€å
        self.ui.horizontalLayout.setContentsMargins(10, 10, 10, 10)
        self.ui.horizontalLayout.setSpacing(10)
        # åˆå§‹åŒ–å˜é‡
        self.history_data = []
        self.setup_connections()
        self.load_history()

    def setup_connections(self):
        self.ui.history_list.itemClicked.connect(self.show_history_detail)
        self.ui.delete_btn.clicked.connect(self.delete_history_item)
        self.ui.export_btn.clicked.connect(self.export_history)
        self.ui.search_input.textChanged.connect(self.search_history)  # æœç´¢åŠŸèƒ½ï¼ˆéœ€åœ¨Designerä¸­æ·»åŠ QLineEditå‘½åä¸ºsearch_inputï¼‰

    def load_history(self):
        """åŠ è½½å†å²è®°å½•"""
        self.ui.history_list.clear()
        self.history_data = []
        try:
            if os.path.exists("history.json"):
                with open("history.json", "r", encoding="utf-8") as f:
                    self.history_data = json.load(f)
            # å€’åºæ˜¾ç¤ºï¼ˆæœ€æ–°çš„è®°å½•åœ¨æœ€ä¸Šé¢ï¼‰
            for item in reversed(self.history_data):
                item_type = "ğŸ“ å¯¹è¯" if item["type"] == "chat" else "ğŸ–¼ï¸ å›¾åƒ"
                self.ui.history_list.addItem(f"{item_type} | {item['time']}")
        except Exception as e:
            self.ui.history_preview.setText(f"åŠ è½½å†å²å¤±è´¥ï¼š{str(e)}")

    def show_history_detail(self, item):
        """æ˜¾ç¤ºé€‰ä¸­è®°å½•çš„è¯¦æƒ…"""
        # è·å–é€‰ä¸­è®°å½•çš„ç´¢å¼•ï¼ˆå€’åºæ˜¾ç¤ºï¼Œéœ€åå‘è®¡ç®—ï¼‰
        index = len(self.history_data) - 1 - self.ui.history_list.row(item)
        if 0 <= index < len(self.history_data):
            data = self.history_data[index]
            # æ ¼å¼åŒ–æ˜¾ç¤ºè¯¦æƒ…
            detail = f"""
            <b>æ—¶é—´ï¼š</b>{data['time']}
            <br><b>ç±»å‹ï¼š</b>{'å¯¹è¯è®°å½•' if data['type'] == 'chat' else 'å›¾åƒç”Ÿæˆè®°å½•'}
            <br><b>å†…å®¹ï¼š</b>
            <br><pre style='background-color:#f5f5f5; padding:8px; border-radius:4px;'>{data['content']}</pre>
            """
            self.ui.history_preview.setHtml(detail)
            # è‹¥ä¸ºå›¾åƒè®°å½•ï¼Œå°è¯•æ˜¾ç¤ºç¼©ç•¥å›¾ï¼ˆéœ€åœ¨Designerä¸­æ·»åŠ QLabelå‘½åä¸ºimage_thumbnailï¼‰
            if data["type"] == "image" and "è·¯å¾„ï¼š" in data["content"]:
                # æå–å›¾åƒè·¯å¾„
                path_line = [line for line in data["content"].split("\n") if "è·¯å¾„ï¼š" in line][0]
                image_path = path_line.split("ï¼š")[-1].strip()
                if os.path.exists(image_path):
                    thumbnail = QPixmap(image_path).scaled(
                        self.ui.image_thumbnail.size(),
                        Qt.KeepAspectRatio,
                        Qt.SmoothTransformation
                    )
                    self.ui.image_thumbnail.setPixmap(thumbnail)
                else:
                    self.ui.image_thumbnail.setText("å›¾åƒæ–‡ä»¶å·²åˆ é™¤")
            else:
                self.ui.image_thumbnail.clear()

    def delete_history_item(self):
        """åˆ é™¤é€‰ä¸­çš„å†å²è®°å½•"""
        current_item = self.ui.history_list.currentItem()
        if not current_item:
            QMessageBox.warning(self, "åˆ é™¤æç¤º", "è¯·å…ˆé€‰ä¸­è¦åˆ é™¤çš„å†å²è®°å½•ï¼")
            return

        # ç¡®è®¤åˆ é™¤
        confirm = QMessageBox.question(
            self, "åˆ é™¤ç¡®è®¤", "ç¡®å®šè¦åˆ é™¤è¿™æ¡å†å²è®°å½•å—ï¼Ÿ",
            QMessageBox.Yes | QMessageBox.No, QMessageBox.No
        )
        if confirm != QMessageBox.Yes:
            return

        # åˆ é™¤æ•°æ®å¹¶æ›´æ–°æ–‡ä»¶
        index = len(self.history_data) - 1 - self.ui.history_list.row(current_item)
        self.history_data.pop(index)
        with open("history.json", "w", encoding="utf-8") as f:
            json.dump(self.history_data, f, ensure_ascii=False, indent=2)
        # é‡æ–°åŠ è½½æ˜¾ç¤º
        self.load_history()
        self.ui.history_preview.clear()
        self.ui.image_thumbnail.clear()
        QMessageBox.information(self, "åˆ é™¤æˆåŠŸ", "é€‰ä¸­çš„å†å²è®°å½•å·²åˆ é™¤ï¼")

    def export_history(self):
        """å¯¼å‡ºæ‰€æœ‰å†å²è®°å½•"""
        if not self.history_data:
            QMessageBox.warning(self, "å¯¼å‡ºæç¤º", "æš‚æ— å†å²è®°å½•å¯å¯¼å‡ºï¼")
            return

        # å¼¹å‡ºå¯¼å‡ºå¯¹è¯æ¡†
        file_path, _ = QFileDialog.getSaveFileName(
            self, "å¯¼å‡ºå†å²è®°å½•",
            f"ai_history_{self.get_current_time()}.txt",
            "æ–‡æœ¬æ–‡ä»¶ (*.txt);;JSONæ–‡ä»¶ (*.json);;æ‰€æœ‰æ–‡ä»¶ (*.*)"
        )
        if not file_path:
            return

        try:
            if file_path.endswith(".json"):
                # å¯¼å‡ºä¸ºJSONæ ¼å¼ï¼ˆä¿ç•™åŸå§‹ç»“æ„ï¼‰
                with open(file_path, "w", encoding="utf-8") as f:
                    json.dump(self.history_data, f, ensure_ascii=False, indent=2)
            else:
                # å¯¼å‡ºä¸ºå¯è¯»æ–‡æœ¬æ ¼å¼
                with open(file_path, "w", encoding="utf-8") as f:
                    f.write(f"AIåŠ©æ‰‹å†å²è®°å½•ï¼ˆå¯¼å‡ºæ—¶é—´ï¼š{self.get_current_time()}ï¼‰\n")
                    f.write("=" * 50 + "\n\n")
                    for i, item in enumerate(reversed(self.history_data), 1):
                        f.write(f"ã€{i}ã€‘{item['time']} | {'å¯¹è¯' if item['type'] == 'chat' else 'å›¾åƒ'}\n")
                        f.write(f"å†…å®¹ï¼š\n{item['content']}\n")
                        f.write("-" * 30 + "\n\n")
            QMessageBox.information(self, "å¯¼å‡ºæˆåŠŸ", f"å†å²è®°å½•å·²å¯¼å‡ºåˆ°ï¼š\n{file_path}")
        except Exception as e:
            QMessageBox.error(self, "å¯¼å‡ºå¤±è´¥", f"å¯¼å‡ºå†å²è®°å½•å‡ºé”™ï¼š{str(e)}")

    def search_history(self, search_text):
        """æœç´¢å†å²è®°å½•ï¼ˆæŒ‰å…³é”®è¯åŒ¹é…ï¼‰"""
        if not search_text:
            # æœç´¢ä¸ºç©ºï¼Œé‡æ–°åŠ è½½æ‰€æœ‰è®°å½•
            self.load_history()
            return

        # ç­›é€‰åŒ…å«å…³é”®è¯çš„è®°å½•
        filtered_data = [
            item for item in self.history_data
            if search_text.lower() in item["content"].lower() or
               search_text.lower() in item["time"].lower()
        ]
        # æ›´æ–°åˆ—è¡¨æ˜¾ç¤º
        self.ui.history_list.clear()
        for item in reversed(filtered_data):
            item_type = "ğŸ“ å¯¹è¯" if item["type"] == "chat" else "ğŸ–¼ï¸ å›¾åƒ"
            self.ui.history_list.addItem(f"{item_type} | {item['time']}")

    @staticmethod
    def get_current_time():
        """è·å–å½“å‰æ—¶é—´ï¼ˆæ ¼å¼ï¼š20240520153045ï¼‰"""
        return datetime.now().strftime("%Y%m%d%H%M%S")



class MainApplication(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("AIåˆ›ä½œåŠ©æ‰‹")
        self.setGeometry(100, 100, 1200, 800)  # å¢å¤§é»˜è®¤çª—å£å°ºå¯¸ï¼Œæå‡ä½“éªŒ
        
        # 1. é…ç½®QTabWidgetä¸ºâ€œè‡ªé€‚åº”æ‰©å±•â€
        self.tabs = QTabWidget()
        self.tabs.setSizePolicy(
            QSizePolicy.Expanding,  # æ°´å¹³æ–¹å‘ï¼šå æ»¡å¯ç”¨ç©ºé—´
            QSizePolicy.Expanding   # å‚ç›´æ–¹å‘ï¼šå æ»¡å¯ç”¨ç©ºé—´
        )
        self.setCentralWidget(self.tabs)
        
        # 2. åˆå§‹åŒ–å­çª—å£ï¼ˆå¯¹è¯ã€å›¾åƒã€å†å²ï¼‰
        self.chat_interface = ChatInterface()
        self.image_generator = ImageGenerator()
        self.history_manager = HistoryManager()
        
        # 3. å…³é”®ï¼šå°†å­çª—å£çš„ä¸­å¤®éƒ¨ä»¶ä½œä¸ºæ ‡ç­¾é¡µå†…å®¹ï¼ˆè€Œéç›´æ¥ç”¨QMainWindowï¼‰
        # åŸå› ï¼šQMainWindowåµŒå¥—æ—¶ï¼Œç›´æ¥æ·»åŠ ä¼šä¿ç•™å…¶è¾¹æ¡†/ç•™ç™½ï¼Œç”¨ä¸­å¤®éƒ¨ä»¶å¯æ¶ˆé™¤
        self.tabs.addTab(self.chat_interface.centralWidget(), "å¯¹è¯åŠ©æ‰‹")
        self.tabs.addTab(self.image_generator.centralWidget(), "å›¾åƒç”Ÿæˆ")
        self.tabs.addTab(self.history_manager.centralWidget(), "å†å²è®°å½•")

if __name__ == "__main__":


    load_dotenv()


    app = QApplication(sys.argv)
    # è®¾ç½®ä¸­æ–‡å­—ä½“
    font = QFont("SimHei")
    app.setFont(font)
    window = MainApplication()
    window.show()
    sys.exit(app.exec_())
